REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
import json

In [3]:
import numpy as np

In [4]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [5]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [6]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [7]:
len(bus.index)

6315

#### Question 2b: How many unique business IDs are there?  

In [8]:
for col in bus:
  print(bus[col].unique())


[   10    19    24 ... 90237 90238 90239]
['TIRAMISU KITCHEN' 'NRGIZE LIFESTYLE CAFE'
 'OMNI S.F. HOTEL - 2ND FLOOR PANTRY' ... "WIN'S RESTAURANT"
 'USA POWER MARKET' "ALBA RAY'S"]
['033 BELDEN PL' '1200 VAN NESS AVE, 3RD FLOOR'
 '500 CALIFORNIA ST, 2ND  FLOOR' ... '3040 TARAVAL ST ' '1524 OCEAN AVE '
 '2293 MISSION ST ']
['San Francisco']
['CA']
['94104' '94109' '94133' '94110' '94122' '94103' '94115' '94131' '94111'
 '94117' '94107' '94108' '94102' '94132' '94105' '94134' '94124' '94116'
 '94121' '94118' '94112' '94123' '94114' '94127' '941102019' '941' '94545'
 '94066' '941033148' nan '94158' '95105' '94013' '94130' 'CA' '92672'
 '94120' '94143' '94609' '94101' '00000' '94188' '94621' '94014' '94129'
 '94602' 'Ca']
[37.791116 37.786848 37.792888 ... 37.798078 37.731568 37.794063]
[-122.403816 -122.421547 -122.403135 ... -122.435444 -122.45097
 -122.400726]
['+14154217044' '+14157763262' '+14156779494' ... '+14154945048'
 '+14158606036' '+14153331878']


In [9]:
x=bus.business_id.value_counts()
x

10       1
75166    1
75139    1
75123    1
75113    1
        ..
19131    1
19109    1
19094    1
19087    1
90239    1
Name: business_id, Length: 6315, dtype: int64

In [10]:
len(x)

6315

In [11]:
n = bus.nunique(axis=0)
  
print("No.of.unique values in each column :\n",
      n)

No.of.unique values in each column :
 business_id     6315
name            5692
address         5534
city               1
state              1
postal_code       46
latitude        2865
longitude       2914
phone_number    4919
dtype: int64


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [12]:
y=bus.name.value_counts(ascending=False)
y

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
                              ..
7-MISSION RESTAURANT           1
WONDERFUL DESSERT & CAFE       1
KITCHEN ISTANBAL               1
FAT BELLI DELI                 1
ALBA RAY'S                     1
Name: name, Length: 5692, dtype: int64

In [13]:
a=bus['name'].value_counts().to_frame()
a.head(5)

name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11

In [14]:
a=json.loads(bus['name'].value_counts().to_json())
a


{'STARBUCKS COFFEE': 72,
 "PEET'S COFFEE & TEA": 24,
 'MCDONALDS': 12,
 'SAN FRANCISCO SOUP COMPANY': 11,
 'WALGREENS': 11,
 "LEE'S DELI": 10,
 'JAMBA JUICE': 9,
 'AT&T PARK - FOOD CART': 9,
 "SPECIALTY'S CAFE & BAKERY": 9,
 'SUBWAY SANDWICHES': 8,
 'BON APPETIT @ TWITTER': 7,
 'PHILZ COFFEE': 7,
 'WHOLE FOODS MARKET': 7,
 'BON APPETIT MANAGEMENT CO': 7,
 'PROPER FOOD': 6,
 'NUCHA': 6,
 'MIXT GREENS': 6,
 'AT&T PARK - DRAFT BEER CART': 6,
 'BLUE BOTTLE COFFEE': 6,
 'SUBWAY': 6,
 'QUICKLY': 5,
 'ROUND TABLE PIZZA': 5,
 'LA BOULANGERIE DE SAN FRANCISCO': 5,
 'JW MARRIOTT SF UNION SQUARE': 5,
 "ANNIE'S HOT DOGS & PRETZELS": 5,
 'ESCAPE FROM NEW YORK PIZZA': 5,
 'THE PLANT CAFE ORGANIC': 5,
 "PHIL'Z COFFEE": 4,
 'MARLA BAKERY': 4,
 'THE MELT': 4,
 'SUSHIRRITO': 4,
 'SUPER DUPER': 4,
 'THE CRÈME BRULEE CART': 4,
 'COASTSIDE FARMS AND SPECIALTIES': 4,
 'EAST & WEST GOURMET FOOD': 4,
 'FRESCA RESTAURANT': 4,
 'HAPPY DONUTS': 4,
 'KENTUCKY FRIED CHICKEN': 4,
 'CURRY UP NOW': 4,
 'PANDA EXPRESS

In [15]:
bus.head(5)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [16]:
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

#### Question 3b

What are the unique values of postal_code?

In [17]:
bus.postal_code.value_counts()

94110        585
94103        558
94102        458
94107        451
94133        417
94109        383
94111        282
94122        270
94118        243
94115        238
94114        230
94105        226
94108        224
94117        197
94112        190
94123        169
94124        164
94121        161
94132        134
94104        133
94116         93
94134         74
94127         71
94131         48
94158         17
94130          8
94143          5
CA             3
Ca             3
94101          2
94013          2
94129          2
94014          2
941033148      1
94602          1
94621          1
94188          1
00000          1
94609          1
95105          1
941            1
94120          1
941102019      1
94545          1
94066          1
92672          1
Name: postal_code, dtype: int64

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [18]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [19]:
vb = bus[bus['postal_code'].isin(validZip)]

In [20]:
len(vb.name.value_counts())

5438

In [21]:
vb.nunique(axis=0)


business_id     5999
name            5438
address         5345
city               1
state              1
postal_code       24
latitude        2808
longitude       2854
phone_number    4737
dtype: int64

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [22]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [96]:
vb_clean = bus[bus['postal_code'].isin(validZip)]

In [24]:
vb_clean

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

In [92]:
vb['not_null_lon'] = np.where(vb['longitude'].notnull(), 1, 0)
vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)
vb

C:\Users\eugen\AppData\Local\Temp\ipykernel_23672\3098755437.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['not_null_lon'] = np.where(vb['longitude'].notnull(), 1, 0)
C:\Users\eugen\AppData\Local\Temp\ipykernel_23672\3098755437.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vb['null_lon'] = np.where(vb['longitude'].isnull(), 1, 0)


business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  not_null_lon  null_lon  
0     37.791116 -122.403816  +14154217044             1         0  
1     37.786848 -122.421547  +14157763262             1         0  
2     37.792888 -122.403135  +14156779494             1         0  
3     37.807155 -122.419004           NaN             1         0  
4     37.747114 -122.413641  +14156415051             1         0  
...         ...         ...           ...           ...       ...  
6310        NaN         NaN           NaN             0         1  
6311        NaN         NaN  +14154945048             0         1  
6312        NaN         NaN  +14158606036             0         1  
6313        NaN         NaN  +14153331878             0         1  
6314        NaN         NaN  +14158881393             0         1  

[5999 rows x 11 columns]

In [93]:
#Solution
vb_clean = vb[['postal_code', 'null_lon', 'not_null_lon']].groupby('postal_code').sum()
vb_clean

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

In [100]:
#Trying Alternate
vb_clean_null_lon = pd.DataFrame(vb_clean.longitude.isnull().groupby(vb_clean['postal_code']).sum())
vb_clean_null_lon

longitude
postal_code           
94102              196
94103              252
94104               59
94105              100
94107              247
94108               82
94109              146
94110              251
94111              115
94112               59
94114              108
94115               76
94116               35
94117               67
94118              100
94121               55
94122              123
94123               53
94124               88
94127               28
94131               14
94132               65
94133              135
94134               29

In [99]:
vb_clean_not_null_lon = pd.DataFrame(vb_clean.longitude.notnull().groupby(vb_clean['postal_code']).sum())
vb_clean_not_null_lon

longitude
postal_code           
94102              262
94103              306
94104               74
94105              126
94107              204
94108              142
94109              237
94110              334
94111              167
94112              131
94114              122
94115              162
94116               58
94117              130
94118              143
94121              106
94122              147
94123              116
94124               76
94127               43
94131               34
94132               69
94133              282
94134               45

In [69]:
vb_clean1=pd.merge(vb_clean_null_lon, vb_clean_not_null_lon, left_on='postal_code',right_on='postal_code')
vb_clean1

longitude_x  longitude_y
postal_code                          
94102                196          262
94103                252          306
94104                 59           74
94105                100          126
94107                247          204
94108                 82          142
94109                146          237
94110                251          334
94111                115          167
94112                 59          131
94114                108          122
94115                 76          162
94116                 35           58
94117                 67          130
94118                100          143
94121                 55          106
94122                123          147
94123                 53          116
94124                 88           76
94127                 28           43
94131                 14           34
94132                 65           69
94133                135          282
94134                 29           45

In [67]:
#vb[['postal_code', 'null_lon', 'not_null_lon']].groupby('postal_code')
vb_clean2=pd.merge(vb_clean1, vb_clean_null_lon, left_on='postal_code',right_on='postal_code')
vb_clean2

business_id                                  name  \
0              10                      TIRAMISU KITCHEN   
1              24    OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
2              76    OMNI S.F. HOTEL - 3RD FLOOR PANTRY   
3              77  OMNI S.F. HOTEL - EMPLOYEE CAFETERIA   
4             140                        CAFE MADELEINE   
...           ...                                   ...   
5994        88340                        JOYFUL KITCHEN   
5995        88443                              LEMONADE   
5996        88793                PARADISE PIZZA & PASTA   
5997        89167          MONTEREY PIZZA AND ASIAN ZAP   
5998        90236                        PHO DONG HUONG   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
2      500 CALIFORNIA ST, 3RD FLOOR  San Francisco    CA       94104   
3       500 CALIFORNIA ST, BASEMENT  San Francisco    CA       94104   
4                300 CALIFORNIA ST   San Francisco    CA       94104   
...                             ...            ...   ...         ...   
5994                1919 OCEAN AVE   San Francisco    CA       94127   
5995            16 WEST PORTAL AVE   San Francisco    CA       94127   
5996           393 WEST PORTAL AVE   San Francisco    CA       94127   
5997             599 MONTEREY BLVD   San Francisco    CA       94127   
5998             667 MONTEREY BLVD   San Francisco    CA       94127   

       latitude  longitude_x  phone_number  longitude_y  longitude  
0     37.791116  -122.403816  +14154217044           74         59  
1     37.792888  -122.403135  +14156779494           74         59  
2     37.792888  -122.403135  +14156779494           74         59  
3     37.792888  -122.403135  +14156779494           74         59  
4     37.793268  -122.400323  +14153623332           74         59  
...         ...          ...           ...          ...        ...  
5994        NaN          NaN  +14153371919           43         28  
5995        NaN          NaN  +14153734199           43         28  
5996        NaN          NaN  +14158760345           43         28  
5997        NaN          NaN  +14152265595           43         28  
5998        NaN          NaN  +14154945048           43         28  

[5999 rows x 11 columns]

In [74]:
#Unable to Rename?
vb_clean1.rename(columns={"postal_code": "null_lon", "longitude_y": "non_null_lon"})
vb_clean1

longitude_x  longitude_y
postal_code                          
94102                196          262
94103                252          306
94104                 59           74
94105                100          126
94107                247          204
94108                 82          142
94109                146          237
94110                251          334
94111                115          167
94112                 59          131
94114                108          122
94115                 76          162
94116                 35           58
94117                 67          130
94118                100          143
94121                 55          106
94122                123          147
94123                 53          116
94124                 88           76
94127                 28           43
94131                 14           34
94132                 65           69
94133                135          282
94134                 29           45

In [75]:
vb_clean1

longitude_x  longitude_y
postal_code                          
94102                196          262
94103                252          306
94104                 59           74
94105                100          126
94107                247          204
94108                 82          142
94109                146          237
94110                251          334
94111                115          167
94112                 59          131
94114                108          122
94115                 76          162
94116                 35           58
94117                 67          130
94118                100          143
94121                 55          106
94122                123          147
94123                 53          116
94124                 88           76
94127                 28           43
94131                 14           34
94132                 65           69
94133                135          282
94134                 29           45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [28]:
vb_clean['per_null'] = vb_clean['null_lon']/(vb_clean['null_lon']+vb_clean['not_null_lon'])
vb_clean

null_lon  not_null_lon  per_null
postal_code                                  
94102             262           196  0.572052
94103             306           252  0.548387
94104              74            59  0.556391
94105             126           100  0.557522
94107             204           247  0.452328
94108             142            82  0.633929
94109             237           146  0.618799
94110             334           251  0.570940
94111             167           115  0.592199
94112             131            59  0.689474
94114             122           108  0.530435
94115             162            76  0.680672
94116              58            35  0.623656
94117             130            67  0.659898
94118             143           100  0.588477
94121             106            55  0.658385
94122             147           123  0.544444
94123             116            53  0.686391
94124              76            88  0.463415
94127              43            28  0.605634
94131              34            14  0.708333
94132              69            65  0.514925
94133             282           135  0.676259
94134              45            29  0.608108

In [29]:
vb_clean.sort_values('per_null', ascending=False).head(5)

null_lon  not_null_lon  per_null
postal_code                                  
94131              34            14  0.708333
94112             131            59  0.689474
94123             116            53  0.686391
94115             162            76  0.680672
94133             282           135  0.676259

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [30]:
ins

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

In [31]:
ins.describe

<bound method NDFrame.describe of        business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]>

In [32]:
pd.crosstab(ins['business_id'],columns='count') #appears to have more than one inspection for many busineses

col_0        count
business_id       
10               3
19               3
24               4
31               3
45               3
...            ...
89515            1
89701            1
89790            1
89803            1
89900            1

[5730 rows x 1 columns]

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [33]:
ins.type.value_counts

<bound method IndexOpsMixin.value_counts of 0        routine
1        routine
2        routine
3        routine
4        routine
          ...   
15425    routine
15426    routine
15427    routine
15428    routine
15429    routine
Name: type, Length: 15430, dtype: object>

In [34]:
pd.crosstab(ins['type'],columns='count')

col_0      count
type            
complaint      1
routine    15429

In [35]:
ins2 = ins.drop(ins.loc[ins['type']=='complaint'].index)
pd.crosstab(ins2['type'], columns= 'count')

col_0    count
type          
routine  15429

In [36]:
ins2

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [37]:
ins2.dtypes

business_id     int64
score           int64
date            int64
type           object
dtype: object

In [38]:
#df['points'] = df['points'].astype(str).str.replace('.', '')

date=ins['date'].astype(str).str[:4]
date

0        2016
1        2014
2        2014
3        2016
4        2014
         ... 
15425    2016
15426    2016
15427    2016
15428    2016
15429    2016
Name: date, Length: 15430, dtype: object

In [39]:
#int(str(number)[:2])
ins2['year'] = ins['date'].astype(str).str[:4]
ins2

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [40]:
pd.crosstab(ins2['year'], columns= 'count')

col_0  count
year        
2013      38
2014    5629
2015    3923
2016    5839

In [47]:
ins2.loc[ins2['year']=='2013']

business_id  score      date     type  year
534            695     96  20131231  routine  2013
761           1043     98  20131226  routine  2013
804           1124     92  20131230  routine  2013
1034          1458     72  20131227  routine  2013
1680          2245     91  20131227  routine  2013
1817          2401     78  20131226  routine  2013
2167          2843     92  20131227  routine  2013
2773          3504     90  20131227  routine  2013
3641          4900     96  20131226  routine  2013
3678          4947     96  20131227  routine  2013
3717          4992     96  20131230  routine  2013
3730          5006     96  20131230  routine  2013
3800          5091     96  20131230  routine  2013
5268          6521     98  20131230  routine  2013
5497          7489     68  20131226  routine  2013
5550          7538     96  20131226  routine  2013
5598          7605     90  20131226  routine  2013
5623          7641     92  20131226  routine  2013
5641          7649     90  20131227  routine  2013
5654          7656     98  20131226  routine  2013
5720          7757     90  20131227  routine  2013
5800          8948     94  20131231  routine  2013
5997         10986     96  20131227  routine  2013
6832         28843     90  20131231  routine  2013
6915         32654     71  20131227  routine  2013
8216         59251     68  20131230  routine  2013
8657         61592    100  20131226  routine  2013
9447         65014     92  20131231  routine  2013
9711         65826     98  20131227  routine  2013
9979         66874     73  20131226  routine  2013
10006        66927     96  20131226  routine  2013
10157        67495     98  20131230  routine  2013
10495        68351     83  20131230  routine  2013
10515        68379     94  20131227  routine  2013
10568        68462    100  20131226  routine  2013
10640        68672     75  20131227  routine  2013
11388        70452     98  20131231  routine  2013
12940        77177     98  20131231  routine  2013

In [48]:
#df = df[df.rebounds != 7]

ins3 = ins2[ins2.year !='2013']

In [49]:
ins3 = ins2.drop(ins2.loc[ins2['year']=='2013'].index) 
pd.crosstab(ins3['year'], columns= 'count')

col_0  count
year        
2014    5629
2015    3923
2016    5839

In [50]:
pd.crosstab(ins3['year'], columns= 'count')

col_0  count
year        
2014    5629
2015    3923
2016    5839

In [51]:
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [52]:
ins_2016 = ins3.loc[ins3['year'] == '2016']
ins_2016

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [54]:
bus_ins2016 = pd.merge(bus, ins_2016, how='left', on=['business_id', 'business_id'])
bus_ins2016.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score        date     type  year  
0 -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1 -122.421547  +14157763262   94.0  20160513.0  routine  2016  
2 -122.403135  +14156779494   98.0  20161005.0  routine  2016  
3 -122.403135  +14156779494   96.0  20160311.0  routine  2016  
4 -122.419004           NaN   84.0  20160816.0  routine  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [55]:
bus_ins2016[['score', 'name', 'address']].sort_values('score', ascending=True).head(20)

score                         name              address
4892   52.0      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD 
7324   55.0                    POKI TIME     2101 LOMBARD ST 
6281   55.0                 CRAZY PEPPER   2257 SAN JOSE AVE 
4873   55.0                   UNCLE CAFE       65 WAVERLY PL 
3008   56.0             HING WANG BAKERY        339 JUDAH ST 
5627   56.0                   GOLDEN WOK   295 B ORIZABA AVE 
5383   57.0                   TOP'S CAFE  2460 SAN BRUNO AVE 
6067   57.0    L & G VIETNAMESE SANDWICH         602 EDDY ST 
836    57.0                 RED A BAKERY      634 CLEMENT ST 
4209   57.0      SAN TUNG RESTAURANT LLC      1031 IRVING ST 
549    57.0          MA'S DIM SUM & CAFE      1315 POWELL ST 
4903   57.0  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST 
6645   58.0    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST 
624    59.0          JOANIE'S DINER INC.   1329 COLUMBUS AVE 
6938   59.0         PAK NAM THAI CUISINE       655 LARKIN ST 
3563   59.0                  JADE GARDEN      3608 BALBOA ST 
4234   59.0                        ITHAI         720 POST ST 
6151   60.0          VALLARTA'S TACO BAR     2022 MISSION ST 
4850   60.0             HONG KONG LOUNGE     5322 GEARY BLVD 
5013   60.0                 WOLFES LUNCH        1220 16TH ST

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!